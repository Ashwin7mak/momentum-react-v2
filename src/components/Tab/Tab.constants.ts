const CLASS_PREFIX = 'md-tab';

const DEFAULTS = {
  ACTIVE: false,
  DISABLED: false,
};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
