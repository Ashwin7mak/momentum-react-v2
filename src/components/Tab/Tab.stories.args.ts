import { commonAriaButton, commonStyles } from '../../storybook/helper.stories.argtypes';
import { TAB_CONSTANTS as CONSTANTS } from './';

export default {
  ...commonStyles,
  ...commonAriaButton,
  children: {
    description: 'Provides the child nodes for this element.',
    control: 'text',
    table: {
      type: {
        summary: 'ReactNode',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  active: {
    description: 'Tells whether the tab is currently active or not.',
    control: 'boolean',
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.ACTIVE,
      },
    },
  },
  disabled: {
    description: 'Whether to render the Tab as disabled.',
    control: 'boolean',
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.DISABLED,
      },
    },
  },
  open: {
    description: 'Tells whether the tab is currently open or not.',
    control: 'boolean',
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.OPEN,
      },
    },
  },
};
