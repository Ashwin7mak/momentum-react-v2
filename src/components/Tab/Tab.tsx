import React, { forwardRef, useRef, FC, RefObject } from 'react';
import classnames from 'classnames';
import { useButton } from '@react-aria/button';
import FocusRing from '../FocusRing';

import { DEFAULTS, STYLE } from './Tab.constants';
import { Props } from './Tab.types';
import './Tab.style.scss';
import Text from '../Text';

const Tab: FC<Props> = forwardRef((props: Props, providedRef: RefObject<HTMLButtonElement>) => {
  const { children, className, id, style, active, disabled, open } = props;
  const ref = providedRef || useRef();
  const mutatedProps = {
    ...props,
    isDisabled: props.disabled,
  };

  delete mutatedProps.className;
  delete mutatedProps.disabled;
  delete mutatedProps.id;
  delete mutatedProps.style;

  const { buttonProps } = useButton(mutatedProps, ref);

  return (
    <FocusRing disabled={disabled}>
      <button
        className={classnames(STYLE.wrapper, className)}
        {...buttonProps}
        data-active={active || DEFAULTS.ACTIVE}
        data-disabled={disabled || DEFAULTS.DISABLED}
        data-open={open || DEFAULTS.OPEN}
        ref={ref}
        id={id}
        style={style}
      >
        {typeof children === 'string' ? <Text type="subheader-secondary">children</Text> : children}
      </button>
    </FocusRing>
  );
});

Tab.displayName = 'Tab';

export default Tab;
