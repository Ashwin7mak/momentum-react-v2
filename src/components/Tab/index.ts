import { default as Tab } from './Tab';
import * as CONSTANTS from './Tab.constants';
import { Props } from './Tab.types';

export { CONSTANTS as TAB_CONSTANTS };

export type TabProps = Props;

export default Tab;
