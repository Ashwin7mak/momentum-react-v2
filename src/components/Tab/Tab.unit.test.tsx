import React from 'react';
import { mount } from 'enzyme';

import Tab, { TAB_CONSTANTS as CONSTANTS } from './';
const { DEFAULTS } = CONSTANTS;

describe('<Tab />', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mount(<Tab>Example label</Tab>);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', () => {
      expect.assertions(1);

      const className = 'example-class';

      const container = mount(<Tab className={className} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', () => {
      expect.assertions(1);

      const id = 'example-id';

      const container = mount(<Tab id={id} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mount(<Tab style={style} />);

      expect(container).toMatchSnapshot();
    });
    it('should match snapshot with active', () => {
      expect.assertions(1);

      const active = DEFAULTS.ACTIVE;

      const container = mount(<Tab active={active} />);

      expect(container).toMatchSnapshot();
    });
    it('should match snapshot with disabled', () => {
      expect.assertions(1);

      const disabled = DEFAULTS.DISABLED;

      const container = mount(<Tab disabled={disabled} />);

      expect(container).toMatchSnapshot();
    });
    it('should match snapshot with open', () => {
      expect.assertions(1);

      const open = DEFAULTS.OPEN;

      const container = mount(<Tab open={open} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its wrapper class', () => {
      expect.assertions(1);

      const element = mount(<Tab />)
        .find(Tab)
        .getDOMNode();

      expect(element.classList.contains(CONSTANTS.STYLE.wrapper)).toBe(true);
    });

    it('should have provided class when className is provided', () => {
      expect.assertions(1);

      const className = 'example-class';

      const element = mount(<Tab className={className} />)
        .find(Tab)
        .getDOMNode();

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should have provided id when id is provided', () => {
      expect.assertions(1);

      const id = 'example-id';

      const element = mount(<Tab id={id} />)
        .find(Tab)
        .getDOMNode();

      expect(element.id).toBe(id);
    });

    it('should have provided style when style is provided', () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';

      const element = mount(<Tab style={style} />)
        .find(Tab)
        .getDOMNode();

      expect(element.getAttribute('style')).toBe(styleString);
    });

    it('should have provided id when active id is provided', () => {
      expect.assertions(1);

      const active = DEFAULTS.ACTIVE;

      const element = mount(<Tab active={active} />)
        .find(Tab)
        .getDOMNode();

      expect(element.getAttribute('data-active')).toBe(`${active}`);
    });

    it('should have provided disabled when disabled disabled provided', () => {
      expect.assertions(1);

      const disabled = DEFAULTS.DISABLED;

      const element = mount(<Tab disabled={disabled} />)
        .find(Tab)
        .getDOMNode();

      expect(element.getAttribute('data-disabled')).toBe(`${disabled}`);
    });

    it('should have provided open when open is provided', () => {
      expect.assertions(1);

      const open = DEFAULTS.OPEN;

      const element = mount(<Tab active={open} />)
        .find(Tab)
        .getDOMNode();

      expect(element.getAttribute('data-open')).toBe(`${open}`);
    });
  });

  describe('actions', () => {
    it('should handle mouse press events', () => {
      expect.assertions(1);

      const mockCallback = jest.fn();

      const component = mount(<Tab onPress={mockCallback} />).find(Tab);

      component.props().onPress({
        type: 'press',
        pointerType: 'mouse',
        shiftKey: false,
        ctrlKey: false,
        metaKey: false,
        target: component.getDOMNode(),
      });

      expect(mockCallback).toBeCalledTimes(1);
    });
  });
});
