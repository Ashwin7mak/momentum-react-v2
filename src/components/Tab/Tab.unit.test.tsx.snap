// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab /> snapshot should match snapshot 1`] = `
<Tab>
  <FocusRing>
    <FocusRing
      focusClass="children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <button
        className="md-tab-wrapper"
        data-active={false}
        data-disabled={false}
        data-open={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        type="button"
      >
        <Text
          type="subheader-secondary"
        >
          <h3
            className="md-text-wrapper md-text-color"
            data-type="subheader-secondary"
          >
            children
          </h3>
        </Text>
      </button>
    </FocusRing>
  </FocusRing>
</Tab>
`;

exports[`<Tab /> snapshot should match snapshot with active 1`] = `
<Tab
  active={false}
>
  <FocusRing>
    <FocusRing
      focusClass="children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <button
        className="md-tab-wrapper"
        data-active={false}
        data-disabled={false}
        data-open={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        type="button"
      />
    </FocusRing>
  </FocusRing>
</Tab>
`;

exports[`<Tab /> snapshot should match snapshot with className 1`] = `
<Tab
  className="example-class"
>
  <FocusRing>
    <FocusRing
      focusClass="children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <button
        className="md-tab-wrapper example-class"
        data-active={false}
        data-disabled={false}
        data-open={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        type="button"
      />
    </FocusRing>
  </FocusRing>
</Tab>
`;

exports[`<Tab /> snapshot should match snapshot with disabled 1`] = `
<Tab
  disabled={false}
>
  <FocusRing
    disabled={false}
  >
    <FocusRing
      disabled={false}
      focusClass="children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <button
        className="md-tab-wrapper"
        data-active={false}
        data-disabled={false}
        data-open={false}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        type="button"
      />
    </FocusRing>
  </FocusRing>
</Tab>
`;

exports[`<Tab /> snapshot should match snapshot with id 1`] = `
<Tab
  id="example-id"
>
  <FocusRing>
    <FocusRing
      focusClass="children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <button
        className="md-tab-wrapper"
        data-active={false}
        data-disabled={false}
        data-open={false}
        id="example-id"
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        type="button"
      />
    </FocusRing>
  </FocusRing>
</Tab>
`;

exports[`<Tab /> snapshot should match snapshot with open 1`] = `
<Tab
  open={false}
>
  <FocusRing>
    <FocusRing
      focusClass="children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <button
        className="md-tab-wrapper"
        data-active={false}
        data-disabled={false}
        data-open={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        type="button"
      />
    </FocusRing>
  </FocusRing>
</Tab>
`;

exports[`<Tab /> snapshot should match snapshot with style 1`] = `
<Tab
  style={
    Object {
      "color": "pink",
    }
  }
>
  <FocusRing>
    <FocusRing
      focusClass="children"
      focusRingClass="md-focus-ring-wrapper children"
    >
      <button
        className="md-tab-wrapper"
        data-active={false}
        data-disabled={false}
        data-open={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragStart={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchCancel={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "color": "pink",
          }
        }
        type="button"
      />
    </FocusRing>
  </FocusRing>
</Tab>
`;
